processor/src/main/java/org/mapstruct/ap/internal/model/common/Type.java

```diff
@@ -495,18 +501,54 @@ else if ( Executables.isFieldAccessor( candidate ) && ( Executables.isFinal( can
             }
 
             Accessor previousCandidate = result.get( targetPropertyName );
-            if ( previousCandidate == null || preferredType == null || ( targetType != null
-                && typeUtils.isAssignable( preferredType.getTypeMirror(), targetType.getTypeMirror() ) ) ) {
+
+
+            if (previousCandidate != null
+                    && sourceAccessor != null
+                    && sourceReadAccessor != null
+                    && !isBetter( previousCandidate, candidate,
+                    sourceReadAccessor.getAccessedType() ) ) {
+                continue;
+            }
+
+
+            if (previousCandidate == null || preferredType == null || (targetType != null
+                    && typeUtils.isAssignable( preferredType, targetType.getTypeMirror() ) ) ) {
                 result.put( targetPropertyName, candidate );
             }
         }
 
         return result;
     }
 
-    private Type determinePreferredType(Accessor readAccessor) {
+    private TypeMirror extractTypeMirror( Accessor accessor ) {
+        return accessor.getExecutable().getParameters().get( 0 ).asType();
```

Comment:
This is wrong it can lead to NPE. The `accessor` does not always have an executable. MapStruct can also map public field and that is why we have introduced the `Accessor`.  I think that this can be done by only using the MapStruct `Type` and not the `TypeMirror`. Maybe you can also use `Type` in the `isBetter` check as well.